rbd-mirror: schedule rebalancer to level-load instances

vshankar opened the pull request.

vshankar commented on src/common/options.cc:
```diff
@@ -6507,6 +6507,15 @@ static std::vector<Option> get_rbd_mirror_options() {
     .set_min(1)
     .set_description("interval (in seconds) to throttle images for mirror daemon peer updates"),
+
+    Option("rbd_mirror_image_policy_rebalance_interval", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
```
We could, but wouldn't it be better to decouple the setting with update interval (which is 1 second by default)? Should the rebalance check run this often?

dillaman replied:
I worry about too many knobs -- so perhaps combine ```rbd_mirror_image_policy_rebalance_on_idle_timeout``` w/ this one

vshankar replied:
We can combine `rbd_mirror_image_policy_rebalance_on_idle_timeout` for both and default to 60 second interval checks if the above config value is zero (we would still need the rebalance timer running right)?

tchaikov commented:
retest this please.

vshankar commented:
@dillaman pr updated

Now the rebalancer task is scheduled only if `rbd_mirror_image_policy_rebalance_on_idle_timeout` is non-zero. A side effect of this change is that once rebalance is enabled (by setting `rbd_mirror_image_policy_rebalance_on_idle_timeout`), the initial rebalance task is only scheduled when one of `ImageMap::update_*` method call is invoked.

dillaman commented on src/common/options.cc:
```diff
@@ -6507,6 +6507,10 @@ static std::vector<Option> get_rbd_mirror_options() {
     .set_min(1)
     .set_description("interval (in seconds) to throttle images for mirror daemon peer updates"),
+
+    Option("rbd_mirror_image_policy_rebalance_on_idle_timeout", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
```
Nit: I would drop the ```_on_idle``` part and just include that in the description.

dillaman commented on src/tools/rbd_mirror/ImageMap.cc:
```diff
@@ -225,6 +238,65 @@ void ImageMap<I>::schedule_update_task() {
+  double resched_after =
+    cct->_conf->get_val<double>("rbd_mirror_image_policy_rebalance_on_idle_timeout");
+
+  if (!resched_after || m_rebalance_task != nullptr) {
```
Nit: you could just cancel any previously scheduled event re-schedule a new one. Then you could drop the ```m_last_policy_update_time``` logic.

vshankar replied:
Wouldn't that be too many cancellations and reschedules as `schedule_rebalance_task()` gets called from `schedule_update_task()`?

dillaman replied:
I wouldn't think it would be a CPU performance concern. Alternatively, you could call   ```schedule_rebalance_task``` only if a new update task was scheduled (i.e. every second).

vshankar replied:
OK -- I kept it simple as of now and invoke `schedule_rebalance_task()` early on from `schedule_update_task()`

dillaman commented on src/tools/rbd_mirror/ImageMap.cc:
```diff
@@ -225,6 +238,65 @@ void ImageMap<I>::schedule_update_task() {
+  // fetch the updated value of idle timeout for (re)scheduling
+  CephContext *cct = reinterpret_cast<CephContext *>(m_ioctx.cct());
+  double resched_after =
+    cct->_conf->get_val<double>("rbd_mirror_image_policy_rebalance_on_idle_timeout");
```
Nit: use ```g_ceph_context``` for config items (can you fix the other instances of ```cct->_conf``` in this file as well?)

vshankar replied:
I'll submit that as a separate pr if that's ok?

dillaman replied:
Sure

vshankar commented:
I also found a bug when am image is undergoing disassociation (due to removal) -- updated the fix in the pr itself as without that images were not getting selected for shuffling upon rebalance.

dillaman commented:
@vshankar Does the new test case also cover the "I also found a bug when am image is undergoing disassociation" bug?

vshankar commented:
> @vshankar Does the new test case also cover the "I also found a bug when am image is undergoing disassociation" bug?

Yes it does -- that gets tested under the hood...

vshankar commented:
pushed a minor cleanup in the unit test

vshankar commented:
retest this please

dillaman commented:
@vshankar Looks like there a memory leak: http://qa-proxy.ceph.com/teuthology/jdillaman-2018-06-08_13:47:08-rbd-wip-jd-testing-distro-basic-smithi/2641914/teuthology.log

vshankar commented:
> @vshankar Looks like there a memory leak: http://qa-proxy.ceph.com/teuthology/jdillaman-2018-06-08_13:47:08-rbd-wip-jd-testing-distro-basic-smithi/2641914/teuthology.log

fixed and updated. issue was in accessing `image_state` reference after it got purged in `execute_policy_action()`. this was not seen earlier as the purge logic used to never get executed.
